bokehjs/test/models/mappers/categorical_marker_mapper.ts

```diff
@@ -0,0 +1,360 @@
+          })
+
+          const vals0 = cm0.v_compute([["a", "1", "foo"]])
+          expect(vals0).to.be.deep.equal(["circle"])
+
+          const vals1 = cm0.v_compute([["a", "1", "baz"]])
+          expect(vals1).to.be.deep.equal(["circle"])
+
+          const cm1 = new CategoricalMarkerMapper({
+            markers,
+            factors: factors.map((x) => x.slice(1, 2)[0]),
+            start: i,
+            end: j,
+          })
+
+          const vals2 = cm1.v_compute([["a", "1", "foo"]])
+          expect(vals2).to.be.deep.equal(["circle"])
+
+          const vals3 = cm1.v_compute([["a", "1", "baz"]])
+          expect(vals3).to.be.deep.equal(["circle"])
+
+          const cm2 = new CategoricalMarkerMapper({
+            markers,
+            factors: factors.map((x) => x.slice(2, 3)[0]),
+            start: i,
+            end: j,
+          })
+
+          const vals4 = cm2.v_compute([["a", "1", "foo"]])
+          expect(vals4).to.be.deep.equal(["circle"])
+
+          const vals5 = cm2.v_compute([["a", "1", "baz"]])
+          expect(vals5).to.be.deep.equal(["circle"])
+        })
+      }
+    })
+
+    describe("and 2-level markers factors", () => {
+
+      it("should map factors to markers with start=0, end=2", () => {
+        const markers: Markers = ["hex", "circle", "dash", "square"]
+        const factors: F3[] = [["a", "1", "foo"], ["a", "2", "foo"], ["b", "2", "baz"], ["c", "1", "bar"]]
+        const cm = new CategoricalMarkerMapper({
+          markers,
+          factors: factors.map((x) => x.slice(0, 2) as F2),
+          end: 2,
+        })
+        const vals = cm.v_compute(factors)
+        expect(vals).to.be.deep.equal(markers)
+      })
+
+      it("should map factors to markers with start=1, end=3", () => {
+        const markers: Markers = ["hex", "circle", "dash", "square"]
+        const factors: F3[] = [["a", "1", "foo"], ["a", "2", "foo"], ["b", "2", "baz"], ["c", "1", "bar"]]
+        const cm = new CategoricalMarkerMapper({
+          markers,
+          factors: factors.map((x) => x.slice(1, 3) as F2),
+          start: 1,
+          end: 3,
+        })
+
+        const vals = cm.v_compute(factors)
+        expect(vals).to.be.deep.equal(markers)
+      })
+
+      for (const [i, j] of [[0, 1], [0, 3], [1, 2], [2, 3]]) {
+        it(`should map everything to default_value with start=${i}, end=${j}`, () => {
+          const markers: Markers = ["hex", "circle", "dash", "square"]
+          const factors: F3[] = [["a", "1", "foo"], ["a", "2", "foo"], ["b", "2", "baz"], ["c", "1", "bar"]]
+
+          const cm0 = new CategoricalMarkerMapper({
+            markers,
+            factors: factors.map((x) => x.slice(0, 2) as F2),
+            start: i,
+            end: j,
+          })
+
+          const vals0 = cm0.v_compute(["a"])
+          expect(vals0).to.be.deep.equal(["circle"])
+
+          /* XXX: gives ["hex"] because original code used [i, j] = [3, 4] due to lack of lexical scoping
```

Comment:
@bryevdv, similarly here, just fails with `["hex"]`, which I think is correct.

Reply:
Yes, i think this specific one should actually map to `hex` as correct behavior